[Site] Improve UX icon search

kbond opened the pull request.

weaverryan committed:
Allowing version 8 of Turbo

jakubtobiasz committed:
[Autocomplete] Allow passing extra options to the autocomplete fields

kbond committed:
minor(live): add security tests

kbond committed:
add initial icon search

kbond committed:
add set selector

kbond committed:
feat(twig): Add attribute rendering system

smnandre committed:
[LiveComponent] Fix component syntax error in doc

1ed committed:
[LiveComponent] Add better error message when hydrating dates

smnandre committed:
[Twig][Live] Drop Twig 2 support

Kocal committed:
[Translator] Update intl-messageformat range version to ^10.5.11, close #1437

weaverryan committed:
[Twig] Removing duplication in docs

kbond committed:
refactor

kbond committed:
add Iconify service

kbond committed:
add icon collection page

kbond committed:
add icon page

kbond committed:
add search to icon collection page

kbond committed:
icon search updates url

kbond committed:
minor: update cs

smnandre committed:
[CI] Upgrade actions/cache@v4

(remove remaining Node16 deprecations)

smnandre committed:
[CI] Fix RST Linter triggers

(avoid running it 40x a day when no doc files have been changed)

kbond committed:
feat(live): improve `TestLiveComponent::actingAs()`

smnandre committed:
Some things

smnandre committed:
Search / icon

weaverryan committed:
[Site] Using URL binding on live components demo

weaverryan committed:
minor #1464 [Site] Using URL binding on live components demo (weaverryan)

This PR was merged into the 2.x branch.

Discussion
----------

[Site] Using URL binding on live components demo

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| Issues        | None
| License       | MIT

Easy win - on https://ux.symfony.com/live-component, as you search in the demo, the URL will now update with the search query. Thanks to that great new feature in 2.14 🔥

Cheers!

Commits
-------

dfe2cebf [Site] Using URL binding on live components demo

kbond commented on ux.symfony.com/src/Iconify.php:
```diff
@@ -36,12 +37,58 @@ public function collection(string $name): ?array
             $item->expiresAfter(604800); // 1 week
 
-            return $this->http->request('GET', 'https://api.iconify.design/collections')
+            $this->http->request('GET', 'https://api.iconify.design/collections')
```
```suggestion
            return $this->http->request('GET', 'https://api.iconify.design/collections')
```

kbond commented:
This is looking really nice, I'm so excited!

kbond commented:
Couple comments:
1. In this PR, do you see a "Allow edits and access to secrets by maintainers" checkbox? If so, feel free to check it so I can push minor tweaks (like the review above) directly to your branch.
2. When clicking the icon in the search results, could we have the icon's page in a dialog/modal?
3. On the icon page, we should have buttons to copy svg, download svg
4. When hovering over the icon in the search results, could we provide quick links for download svg, copy svg, and copy ux:icon:import command? When the "add to cart" functionality is added in the future, we could add this button here as well.

smnandre committed:
Icon

yobrx committed:
Used a Composer class to check for package existence in projects with varying structures

evertharmeling committed:
[Autocomplete] Add support for doctrine/orm:^3.0

weaverryan committed:
minor #1447 [Twig] Removing duplication in docs (weaverryan)

This PR was squashed before being merged into the 2.x branch.

Discussion
----------

[Twig] Removing duplication in docs

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| Issues        | None
| License       | MIT

Somehow, I think during some rewriting I did, I didn't remove old sections. This removes the duplicated sections, but keeps one piece that was new.

Cheers!

Commits
-------

2dcd7b30 [Twig] Removing duplication in docs

weaverryan committed:
feature #1317 [Turbo] Allowing version 8 (weaverryan)

This PR was merged into the 2.x branch.

Discussion
----------

[Turbo] Allowing version 8

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Issues        | None
| License       | MIT

Turbo 8 is now beta! This allows Turbo 8 to be installed. There are no BC breaks that would affect us that i'm aware of, which is not surprising as we do very little with Turbo other than supply it to the user.

Once we merge and release this, users installing TurboBundle will get v8 by default, even while in beta stage. From AssetMapper's perspective, that happens because https://data.jsdelivr.com/v1/packages/npm/`@hotwired`/turbo/resolved already selects beta. I can't find any specifier - e.g. https://data.jsdelivr.com/v1/packages/npm/`@hotwired`/turbo/resolved?specifier=stable that selects the latest, non-beta version.

Cheers!

Commits
-------

7fe845ba Allowing version 8 of Turbo

weaverryan committed:
bug #1467 [StimulusBundle] UxPackageReader class doesn't support projects with varied structures using Composer with custom directory structure (yobrx)

This PR was merged into the 2.x branch.

Discussion
----------

[StimulusBundle] UxPackageReader class doesn't support projects with varied structures using Composer with custom directory structure

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Issues        | no
| License       | MIT

Hello,

I've encountered an issue with the `UxPackageReader` class within the Symfony UX package, where it fails to accurately detect package existence in projects that have a structure differing from the standard Symfony application structure, despite these projects using Composer for package management.

## Description of the Issue
The `UxPackageReader` class appears to be designed with a certain project structure, which works well for typical Symfony applications. However, in projects that deviate from this structure (for instance, microservices or applications with a modular architecture), the class does not effectively locate and read the Composer-managed packages.

## Specific Scenario
To give a concrete example, my project consists of multiple directories at the root, each representing a sub-project within a larger ecosystem. These sub-projects are designed to have their own `vendor` directory for autoloading purposes. However, all Composer packages are installed in a shared `vendor` directory at the root of the project. This setup is beneficial for my project's architecture but unfortunately, `UxPackageReader` is unable to recognize the packages in this unique structure.

### Example :
```

app1
|- vendor
  |- composer
    |- autoload-*.php -> reference files to ../vendor/
    |- ... and other files of composer
app2
|- subapp1
  |- vendor
    |- composer
      |- autoload-*.php -> reference files to ../../vendor/
      |- ... and other files of composer
|- subapp2_legacy
  |- vendor
    |- composer
      |- autoload-*.php -> reference files to ../../vendor/
      |- ... and other files of composer
vendor
 |- symfony
   |- packages
     |- v6.3
       |- _files of package_
     |- v6.4
```

## Proposed Solution
To address this limitation, I have implemented a workaround by utilizing the `InstalledVersions` class from Composer, whenever it is available. This class offers a more robust and flexible way to check for the existence of packages, as it does not rely on the project's directory structure but instead directly reads from Composer's autoload mechanisms and installed packages metadata.

## Suggested Changes
I suggest that the Symfony UX team considers integrating a similar approach into the `UxPackageReader` class. By leveraging `InstalledVersions` or an equivalent method, Symfony UX can enhance its compatibility with a broader range of project structures that use Composer, thus increasing its versatility and usability.

This modification would not only resolve the issue for projects with unique structures but also align the package detection mechanism with Composer's native capabilities, ensuring a more consistent and reliable detection process.

My solution is not testable because it uses Composer class with static methods. Therefore, I was unable to write tests for it. However, I am open to any suggestions for improvement.

Commits
-------

0b175126 Used a Composer class to check for package existence in projects with varying structures

weaverryan committed:
[Turbo] Fixing support for not using old ClassUtils

weaverryan committed:
minor #1453 CS Update (kbond)

This PR was merged into the 2.x branch.

Discussion
----------

CS Update

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| Issues        | n/a
| License       | MIT

Updates to the latest Symfony CS rules.

Refs: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7773 & https://github.com/symfony/symfony/pull/53612

Commits
-------

aad4e17d minor: update cs

weaverryan committed:
bug #1471 [Turbo] Fixing support for not using old ClassUtils (weaverryan)

This PR was squashed before being merged into the 2.x branch.

Discussion
----------

[Turbo] Fixing support for not using old ClassUtils

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| Issues        | Fix #1469
| License       | MIT

Commits
-------

4753fc35 [Turbo] Fixing support for not using old ClassUtils

weaverryan committed:
Bumping 2.14.2 CHANGELOG and reverting minor feature temporarily

weaverryan committed:
minor #1473 Bumping 2.14.2 CHANGELOG and reverting minor feature temporarily (weaverryan)

This PR was merged into the 2.x branch.

Discussion
----------

Bumping 2.14.2 CHANGELOG and reverting minor feature temporarily

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| Issues        | None
| License       | MIT

Hi!

I'd like to get a bug fix release out - especially for the UX turbo package and newer Doctrine versions. Out of an abundance of caution, I'm reverting the Turbo 8 support in this PR, so the patch release is clean. After merging, I'll remake a PR to re-add the support.

Cheers!

Commits
-------

3eb560ac Bumping 2.14.2 CHANGELOG and reverting minor feature temporarily

jrushlow committed:
[Turbo] add missing use statement

weaverryan committed:
bug #1475 [Turbo] add missing use statement (jrushlow)

This PR was merged into the 2.x branch.

Discussion
----------

[Turbo] add missing use statement

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| Issues        |
| License       | MIT

The new `Ux\...\ClassUtil` and the legacy `Doctrine\...\ClassUtils` are very similar. Adding the missing use statement with an alias to avoid IDE confusion :)

Commits
-------

6e5fe193 [Turbo] add missing use statement

weaverryan committed:
[Turbo] Allowing Turbo 8

jrushlow committed:
[Turbo] simplify returning entity class name

weaverryan committed:
feature #1442 [Twig] Add (alternate) attribute rendering system (kbond)

This PR was merged into the 2.x branch.

Discussion
----------

[Twig] Add (alternate) attribute rendering system

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Issues        | n/a
| License       | MIT

This is an alternate to https://github.com/symfony/ux/pull/1404 as suggested by Ryan [here](https://github.com/symfony/ux/pull/1404#issuecomment-1915640350).

This PR allows the following:

```twig
<div
    class="{{ attributes.render('class') }} appended-default"
    style="prepended-default {{ attributes.render('style') }}"
    data-foo="{{ attributes.render('data-foo')|default('replaced-default') }}"
    {{ attributes }}
>
```

When calling `attributes.render('name')` ~(or magically via `attributes.name`)~, the attribute is marked as _rendered_. Later when calling just `{{ attributes }}`, the attributes marked as already rendered are excluded. Whether or not an attribute is considered rendered is only evaluated when converting `ComponentAttributes` to a string.

TODO:
- [x] Docs
- [x] Add test ensuring works in real twig component

Commits
-------

0d027d50 feat(twig): Add attribute rendering system

weaverryan committed:
feature #1322 [Autocomplete] Allow passing extra options to the autocomplete fields (jakubtobiasz)

This PR was squashed before being merged into the 2.x branch.

Discussion
----------

[Autocomplete] Allow passing extra options to the autocomplete fields

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Issues        | n/a
| License       | MIT

This PR still needs some tweaks and tests, but I want to validate my idea. Currently, if we pass anything as a third argument in the `->add()` method, it is only used on the form render, and is fully ignored during the AJAX call.
![CleanShot 2023-12-03 at 12 36 59](https://github.com/symfony/ux/assets/80641364/28bc4b75-215c-4588-80b2-0db9eb82dcc6)

While implementing UX's Autocomplete in Sylius, I wanted to complete the following user story
```
Given I want to edit a taxon
When I want to assign a parent taxon to it
And I check a list of available taxa
Then I should see all taxa except the edited one
```

So basically, I have to pass a **current** taxon's ID to the autocomplete's query. As we know, currently it's not possible. So after contacting `@weaverryan`, I decided to implement a mechanism similar to the one from Live Components.

When you pass an array of options as a `3rd` argument, you can use a special `extra_options` key, which is an array consisting `scalars`/`arrays`/`nulls`.

Next, when the form is rendered, I get these values, calculate a checksum for them, then pass them through `json_encode` and `base64_encode` functions. In the end, I glue them to the `url` values in the `\Symfony\UX\Autocomplete\Form\AutocompleteChoiceTypeExtension::finishView` method.

So, basically with the following configuration:
![CleanShot 2023-12-03 at 12 48 51](https://github.com/symfony/ux/assets/80641364/e7b2585c-f052-4803-9473-a83930a4d1c8)

we end up with the following HTML code:
![CleanShot 2023-12-03 at 12 49 36](https://github.com/symfony/ux/assets/80641364/a173220f-19ed-4104-951c-0fee3469768e)

I decided to "glue" the `extra_options` to the URL, as I didn't have to deal with JS. Of course, I do not exclude a chance to refactor it, as a whole method should be refactored anyway.

Finally, the controller decodes the data, checks the checksum and passes the values to the autocomplete's form.

Commits
-------

308daef1 [Autocomplete] Allow passing extra options to the autocomplete fields

evertharmeling committed:
[StimulusBundle] Set "use_yield" option to fix deprecation message

weaverryan committed:
[Live] Fixed idiomorph id child handling & JavaScript cleanup

weaverryan committed:
bug #1477 [Turbo] simplify returning entity class name (jrushlow)

This PR was merged into the 2.x branch.

Discussion
----------

[Turbo] simplify returning entity class name

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes/no <!-- please update src/**/CHANGELOG.md files -->
| Issues        | Fix #
| License       | MIT

credit goes to nicolas, but alternate way to return the class name

Commits
-------

cded3913 [Turbo] simplify returning entity class name

partulaj committed:
[Docs] Fix a minor typo

weaverryan committed:
minor #1478 [StimulusBundle/TwigComponent] Fix "use_yield" to "true" deprecation notice in CI (evertharmeling)

This PR was merged into the 2.x branch.

Discussion
----------

[StimulusBundle/TwigComponent] Fix "use_yield" to "true" deprecation notice in CI

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| Issues        |
| License       | MIT

Only this one remains (in StimulusBundle), but I can't find a direct implementation related to the `Environment` in that test. Maybe also related to https://github.com/symfony/symfony/pull/53805 or thoughts anyone?

```
1x: Since twig/twig 3.9.0: Not setting "use_yield" to "true" is deprecated.
    1x in StimulusControllerLoaderFunctionalTest::testFullApplicationLoad from Symfony\UX\StimulusBundle\Tests\AssetMapper
```

Commits
-------

1dbb7e0f [StimulusBundle] Set "use_yield" option to fix deprecation message

weaverryan committed:
minor #1455 [CI] Fix RST Linter triggers (smnandre)

This PR was merged into the 2.x branch.

Discussion
----------

[CI] Fix RST Linter triggers

(avoid running it 40x a day when no doc files have been changed)

Commits
-------

a8235fb2 [CI] Fix RST Linter triggers

weaverryan committed:
minor #1460 minor(live): add security tests (kbond)

This PR was merged into the 2.x branch.

Discussion
----------

minor(live): add security tests

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| Issues        | Could help solve/replicate #1124
| License       | MIT

This PR adds some security-related tests:

- Ensure `IsGranted` at the class level is respected.
- Ensure `#[CurrentUser]` can be injected into actions/listeners.

Commits
-------

76c7af55 minor(live): add security tests

smnandre committed:
[Twig][Live] Skip Twig 3.9 🚒  (do not set `use_yield = true` for now)

smnandre commented:
> In this PR, do you see a "Allow edits and access to secrets by maintainers" checkbox? If so, feel free to check it so I can push minor tweaks (like the review above) directly to your branch.

Nope, i can't even close it / set it draft / etc.. as you opened it i guess ?

kbond commented:
Hmm yeah I guess so

smnandre commented:
(yes for 2 / 3 / 4... tomorrow night :) )

smnandre committed:
wip stuff

kbond commented:
Thought of a number 5 for my list above: a way to copy just the icon name (ie `flowbite:user-solid`). I'm imagining a scenario where a user has installed an entire pack into their app. To know what icons they have available, we'll suggest they visit https://ux.symfony.com/icons/flowbite to see the icons they can use and quickly copy the names of ones they want to use in their app. Maybe optionally copy the entire component string: `<twig:UX:Icon name="flowbite-user-solid">`?

This is getting to be a lot of buttons now...

1. Copy name: `flowbite:user-solid`
2. Copy as twig function: `{{ ux_icon('flowbite-user-solid`) }}`
3. Copy as html twig component: `<twig:UX:Icon name="flowbite-user-solid">`
4. Copy import command: `symfony ux:icons:import flowbite:user-solid`
5. Copy raw svg: `<svg...</svg>`
6. Download svg
7. _Add to cart_ (in future)

On the actual icon profile page: https://ux.symfony.com/icons/flowbite/user-solid, we have a lot of real estate, so not an issue showing all these links. When browsing icons, we should tighten these up to just the most useful. Maybe:

1. Copy as twig function
2. Copy as html twig component
3. Copy import command
4. _Add to cart_ (in future)

WDYT?

_My UX designer hat doesn't fit very well so maybe you have better ideas!_

smnandre commented:
For me, we need 2 buttons immediately accessible on the Icon card: 
- copy name (only on the icon name, with a "copy" icon + tooltip "Copy bi:settings"
- add to cart (in future) (only on a + icon on the top right)

Any click elsewhere should open a (big) panel/modal with the other buttons you listed / icon preview (not sure we should keep the icon page for v1.0)

WDYT ?

kbond commented:
Fine by me!

weaverryan committed:
feature #1484 [Live] Fixed idiomorph id child handling & JavaScript cleanup (weaverryan)

This PR was squashed before being merged into the 2.x branch.

Discussion
----------

[Live] Fixed idiomorph id child handling & JavaScript cleanup

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| Issues        |  Fix #1470 Fix #1472
| License       | MIT

Hi!

TODO:

* [x] Finish fixing minor js tests

Sorry, quite a lot going on here to get morphing & child rendering *just* right after switching to Idiomorph. The referenced tickets exposed a few cases that were not handled by the tests:

## IMPORTANT CHANGES

* A) `data-live-id` is now `id`. That's because Idiomorph uses `id` as its special key. So it's time to embrace that.
* B) Fixed a bug if `ChildComponentA` were changed to `ChildComponentB` in the same location. Under some situations, we would NOT swap for the new HTML, and would instead "refresh" `ChildComponentA`.
* C) In the above situation, even when we DID swap to `ChildComponentB` correctly, the `LiveController` would still be attached to a `Component` object representing `ChildComponentA`. That would cause problems later on. We now watch for the `id` attribute to change, and then re-create a new `Component` if changed.
* D) Removed special "morph" handling for children components (which are always empty). Instead added a new `data-live-preserve` attribute for elements that want to keep their current element. Improved handling of this.

## OTHER CHANGES

* Various internal attributes like `data-live-emit` were changed to Stimulus values, like `data-events-to-emit-value`. These are not attributes the user ever sets directly.
* Moved the remaining JS child handling into a plugin
* Added `morph:started` and `morph:finished`
* Simplified "child" handling. Previously we really tried to keep track of our children as they were attached / detached. Now, if we need to know our children, we loop over all components to find children.

Cheers!

Commits
-------

6d593828 [Live] Fixed idiomorph id child handling & JavaScript cleanup

weaverryan committed:
feature #1476 [Turbo] Allowing Turbo 8 (weaverryan)

This PR was merged into the 2.x branch.

Discussion
----------

[Turbo] Allowing Turbo 8

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Issues        | None
| License       | MIT

Turbo 8 is out 🔥 https://github.com/hotwired/turbo/releases/tag/v8.0.0

I'm not aware of any other changes we need to make directly. Everything should work :).

Cheers!

Commits
-------

219bf5a2 [Turbo] Allowing Turbo 8

weaverryan committed:
Bootstrapping an UPGRADE log

weaverryan committed:
[Site] Leveraging Turbo 8 on ux.symfony.com

weaverryan committed:
feature #1486 [Twig][Live] Skip Twig 3.9 🚒  (do not set `use_yield = true` for now) (smnandre)

This PR was squashed before being merged into the 2.x branch.

Discussion
----------

[Twig][Live] Skip Twig 3.9 🚒  (do not set `use_yield = true` for now)

As Symfony Twig Bridge & Twig Bundle have decided to skip Twig 3.9 for now, we should do the same, as testing and/or maintaining a compatible version would be really hard without the bundle support.

https://github.com/symfony/twig-bridge/commit/2abddb106523c53aebfe075c04a9d01bc8cdc5b4
https://github.com/symfony/twig-bundle/commit/23a02ff1693c2c76a4fd55eae83704048d1e93e7

Update:

Twig 3.9 introduce a major change in the way templates are rendered, and to fully test/fix our compatibility, we need to wait Bridge and Bundle to allow Twig 3.9

In the meantime, please do not set "use_yield = true" as we are not ready yet.

This is something that has a major impact on TwigComponent and LiveComponent, and we'll probably need some changes on the CI / test suite to handle both modes.

---

Pull Request (compatibility - wip) : https://github.com/symfony/ux/pull/1487 (help welcomed)
Issue: https://github.com/symfony/ux/issues/1390
Twig PR : https://github.com/twigphp/Twig/pull/3950

Commits
-------

61653842 [Twig][Live] Skip Twig 3.9 🚒  (do not set `use_yield = true` for now)

weaverryan committed:
minor #1485 [Docs] Fix a minor typo (partulaj)

This PR was merged into the 2.x branch.

Discussion
----------

[Docs] Fix a minor typo

Missing ? for property that can be null

Commits
-------

43391d2e [Docs] Fix a minor typo

smnandre committed:
Icon modal ++

smnandre committed:
Build reproductible TemplateMap

weaverryan committed:
[Live] Fixing bug where the active input would maintain its value, but lose its cursor position

weaverryan committed:
[Autocomplete] Fix 2 bugs where TomSelect would reset when not necessary

weaverryan committed:
minor #1495 [Site] Leveraging Turbo 8 on ux.symfony.com (weaverryan)

This PR was merged into the 2.x branch.

Discussion
----------

[Site] Leveraging Turbo 8 on ux.symfony.com

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| Issues        | None
| License       | MIT

That's all that's needed to use Turbo 8 + view transitions + instaclick-style preloading.

https://github.com/symfony/ux/assets/121003/881dbe3e-5e6c-476b-bf2d-3da836501c18

Commits
-------

c5e8d50e [Site] Leveraging Turbo 8 on ux.symfony.com

weaverryan committed:
feature #1468 [Autocomplete] Add support for doctrine/orm:^3.0 (evertharmeling)

This PR was squashed before being merged into the 2.x branch.

Discussion
----------

[Autocomplete] Add support for doctrine/orm:^3.0

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes-ish?
| Issues        | Replaces #1459
| License       | MIT

Follow up on #1459 as it seemed a bit more complicated :)

Because in the current `getPropertyMetadata` function, [see](https://github.com/symfony/ux/blob/2.x/src/Autocomplete/src/Doctrine/EntityMetadata.php#L44) and below:

```php
public function getPropertyMetadata(string $propertyName): array
{
    if (\array_key_exists($propertyName, $this->metadata->fieldMappings)) {
        return $this->metadata->fieldMappings[$propertyName];
    }

    if (\array_key_exists($propertyName, $this->metadata->associationMappings)) {
        return $this->metadata->associationMappings[$propertyName];
    }

    throw new \InvalidArgumentException(sprintf('The "%s" field does not exist in the "%s" entity.', $propertyName, $this->metadata->getName()));
}
```

The function combines getting metadata from `fieldMappings` and `associationMappings`, that both return an `array` in **Doctrine ORM 2**.

This however changes in **Doctrine ORM 3** in `fieldMappings` returning a `FieldMapping` object and `associationMappings` in an `AssociationMapping` object.

To support both Doctrine ORM 2 and 3, I didn't changed the return type of the function for now, as we would otherwise need to bump the major version for the `Autocomplete` component.

To ease the transition to **Doctrine ORM 3** in the future, I've splitted `getPropertyMetadata` in `getFieldMetadata` and `getAssociationMetadata`.

## TODO
- [x] Should we keep the `getPropertyMetadata()` function as a proxy to the new methods? (I removed it for now as the method shouldn't be used directly as [mentioned](https://github.com/symfony/ux/pull/1459#issuecomment-1928120377).
- [x] Should we retroactively add the ``@internal`` the `src/Autocomplete/src/Doctrine/EntityMetadata.php` class also (as [mentioned](https://github.com/symfony/ux/pull/1459#issuecomment-1928120377))?
- [ ] Test on Doctrine ORM 3, currently blocked by https://github.com/zenstruck/foundry/pull/556

Commits
-------

20f4dadc [Autocomplete] Add support for doctrine/orm:^3.0

weaverryan committed:
minor #1494 Bootstrapping an UPGRADE log (weaverryan)

This PR was squashed before being merged into the 2.x branch.

Discussion
----------

Bootstrapping an UPGRADE log

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| Issues        | None
| License       | MIT

As suggested by `@smnandre`, a CHANGELOG. This would be for things that user should actually change from one version to another. symfony/symfony maintains something similar.

Commits
-------

58de823c Bootstrapping an UPGRADE log

smnandre committed:
[Twig][Live] Migrate PHPUnit configuration

weaverryan committed:
bug #1497 Build reproductible TemplateMap (smnandre)

This PR was squashed before being merged into the 2.x branch.

Discussion
----------

Build reproductible TemplateMap

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| Issues        | Fix #1493
| License       | MIT

(prevent crashes when a LiveComponent has to work before/after a cache clear)

See https://github.com/symfony/ux/issues/1493

Commits
-------

5849a6c2 Build reproductible TemplateMap

weaverryan committed:
bug #1501 [Live] Fixing bug where the active input would maintain its value, but lose its cursor position (weaverryan)

This PR was squashed before being merged into the 2.x branch.

Discussion
----------

[Live] Fixing bug where the active input would maintain its value, but lose its cursor position

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| Issues        | Fix #1490
| License       | MIT

This only affects "model" elements, as unmapped element changes are tracked and the "to" element's value updated in the beforeNodeMorphed() callback.

Also, this setting matches Turbo 8. This also updates Idiomorph to the absolute latest version, which includes a related bug fix.

Cheers!

Commits
-------

d3d475aa [Live] Fixing bug where the active input would maintain its value, but lose its cursor position

weaverryan committed:
bug #1502 [Autocomplete] Fix 2 bugs where TomSelect would reset when not necessary (weaverryan)

This PR was merged into the 2.x branch.

Discussion
----------

[Autocomplete] Fix 2 bugs where TomSelect would reset when not necessary

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| Issues        | Fix #1499
| License       | MIT

Hi!

The Autocomplete component "resets" the TomSelect instance under certain situations:

* A) The underlying `<option>` elements have changed
* B) The `<select>` changed its `multiple` attribute

Both of these situations had a bug:

* A) If there was no empty `<option value"">` element, then each time a value was selected, it incorrectly looked like the options had changed, triggering a reset.
* B) If the `<select multiple>` attribute was used without the `multiple="multiple"`, it would incorrectly look like the multiple value was changing, when in fact it was just varying between these 2 valid formats.

In both cases, when combined with LiveComponents, an infinite loop was triggered. That's because, as part of the reset process, when we recreate the TomSelect instance, we "select" the original value. This triggers a `change` event, causing LiveComponents to re-render. On its own , that's fine. But the re-render would trigger one of the bugs above, which would trigger another reset and another re-render.

Cheers!

Commits
-------

7a801fad [Autocomplete] Fix 2 bugs where TomSelect would reset when not necessary

weaverryan committed:
minor #1454 [CI] Upgrade actions/cache@v4 (smnandre)

This PR was merged into the 2.x branch.

Discussion
----------

[CI] Upgrade actions/cache@v4

(remove remaining Node16 deprecations)

Commits
-------

9f052542 [CI] Upgrade actions/cache@v4

weaverryan committed:
minor #1503 [Twig][Live] Migrate PHPUnit configuration (smnandre)

This PR was squashed before being merged into the 2.x branch.

Discussion
----------

[Twig][Live] Migrate PHPUnit configuration

```
Warning:       Your XML configuration validates against a deprecated schema.
Suggestion:    Migrate your XML configuration using "--migrate-configuration"!
```

(will batch the other ones if there are no bad surprises in the CI with those ones)

Commits
-------

238866de [Twig][Live] Migrate PHPUnit configuration

smnandre committed:
[Site] Fix TomSelect select colors in dark mode

(and remove a little closing tag trying to outsmart everyone)

itsdanielfelix committed:
Update index.rst - updated the Livewire website link

Updated the Livewire website link

weaverryan committed:
[Autocomplete] Fix regression where placeholder change didn't cause a reset

smnandre committed:
[Site] Replace stimulus_* helpers with data-attributes

smnandre committed:
Fix charts.js peerDepencies constraint

simonsolutions committed:
Update AutocompleteBundle.de.php

add German translation for string "load more results"

weaverryan committed:
bug #1513 [Autocomplete] Fix regression where placeholder change didn't cause a reset (weaverryan)

This PR was merged into the 2.x branch.

Discussion
----------

[Autocomplete] Fix regression where placeholder change didn't cause a reset

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| Issues        | None
| License       | MIT

Introduced here - https://github.com/symfony/ux/pull/1502/files#diff-655bee3f1f3d92a2b984a8599df80a4f31a842d5f04fd3d186f8e3161ce49c0bR429 - when fixing the bug:

> A) If there was no empty <option value""> element, then each time a value was selected, it incorrectly looked like the options had changed, triggering a reset. This is because TomSelect adds a `<option value="">` is there was not one at the start.

Commits
-------

d81648b8 [Autocomplete] Fix regression where placeholder change didn't cause a reset

weaverryan committed:
minor #1517 [Site] Replace stimulus_* helpers with data-attributes (smnandre)

This PR was squashed before being merged into the 2.x branch.

Discussion
----------

[Site] Replace stimulus_* helpers with data-attributes

Fixes #1514 - cc `@seb`-jean

`@weaverryan`  : I had to adapt a bit the SourceCleaner to keep data-* attributes.

Commits
-------

a32d41e5 [Site] Replace stimulus_* helpers with data-attributes

weaverryan committed:
feature #1518 [Chart.js] Fix peerDepencies constraint (smnandre)

This PR was merged into the 2.x branch.

Discussion
----------

[Chart.js] Fix peerDepencies constraint

Commits
-------

e6686f41 Fix charts.js peerDepencies constraint

weaverryan committed:
feature #1521 Update AutocompleteBundle.de.php (simonsolutions)

This PR was merged into the 2.x branch.

Discussion
----------

Update AutocompleteBundle.de.php

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| Issues        |
| License       | MIT

Add German translation for string "load more results"

Commits
-------

2697fea4 Update AutocompleteBundle.de.php

weaverryan committed:
minor #1510 [Site] Fix TomSelect select colors in dark mode (smnandre)

This PR was merged into the 2.x branch.

Discussion
----------

[Site] Fix TomSelect select colors in dark mode

Fix this bug i had on my todo for a loong time 😶

![image](https://github.com/symfony/ux/assets/1359581/e31ad727-3c3c-4e52-93ba-2086bc55e2be)

(and remove a little closing tag trying to outsmart everyone)

Commits
-------

701720cf [Site] Fix TomSelect select colors in dark mode

weaverryan committed:
2.15 CHANGELOG

weaverryan committed:
minor #1524 2.15 CHANGELOG (weaverryan)

This PR was merged into the 2.x branch.

Discussion
----------

2.15 CHANGELOG

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| Issues        | None
| License       | MIT

Normal releases happen near the end of each month. Due to the morphing library change in 2.14.0, there were enough changes & bugs fixed that I'd like to push an earlier release now.

Cheers!

Commits
-------

bac577c1 2.15 CHANGELOG

weaverryan committed:
minor #1511 Update index.rst - updated the Livewire website link (itsdanielfelix)

This PR was merged into the 2.x branch.

Discussion
----------

Update index.rst - updated the Livewire website link

Updated the Livewire website link

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| Issues        |  -
| License       | MIT

<!--
Replace this notice by a description of your feature/bugfix.
This will help reviewers and should be a good start for the documentation.

Additionally (see https://symfony.com/releases):
 - Always add tests and ensure they pass.
 - For new features, provide some code snippets to help understand usage.
 - Features and deprecations must be submitted against branch main.
 - Changelog entry should follow https://symfony.com/doc/current/contributing/code/conventions.html#writing-a-changelog-entry
 - Never break backward compatibility (see https://symfony.com/bc).
-->

Commits
-------

f48c4a43 Update index.rst - updated the Livewire website link

weaverryan committed:
feature #1461 feat(live): improve `TestLiveComponent::actingAs()` (kbond)

This PR was squashed before being merged into the 2.x branch.

Discussion
----------

feat(live): improve `TestLiveComponent::actingAs()`

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| Issues        | n/a
| License       | MIT

_(depends on #1460)_

Currently, if you have a component that requires authentication for all actions (ie `#[IsGranted]` on the component class), in your tests, as soon as you call `$this->createLiveComponent(...)`, you'll get an access denied exception. You never get a chance to call `->actingAs()`.

This PR makes the initial request for the live component _lazy_. Only when the initial request is required (for other methods), is it made. This gives you the chance to call `->actingAs()` before the initial render.

Commits
-------

d0a4713d feat(live): improve `TestLiveComponent::actingAs()`

weaverryan committed:
feature #1443 [Translator] Update intl-messageformat range version to ^10.5.11, close #1437 (Kocal)

This PR was merged into the 2.x branch.

Discussion
----------

[Translator] Update intl-messageformat range version to ^10.5.11, close #1437

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes <!-- please update src/**/CHANGELOG.md files -->
| Issues        | Fix #1437 <!-- prefix each issue number with "Fix #", no need to create an issue if none exist, explain below instead -->
| License       | MIT

<!--
Replace this notice by a description of your feature/bugfix.
This will help reviewers and should be a good start for the documentation.

Additionally (see https://symfony.com/releases):
 - Always add tests and ensure they pass.
 - For new features, provide some code snippets to help understand usage.
 - Features and deprecations must be submitted against branch main.
 - Changelog entry should follow https://symfony.com/doc/current/contributing/code/conventions.html#writing-a-changelog-entry
 - Never break backward compatibility (see https://symfony.com/bc).
-->

This way, we can ensure users will use a version of `intl-messageformat` that use ``@formatjs`/icu-messageformat-parser`, a 6x faster and backward compatible implementation of `intl-messageformat-parser`.

Commits
-------

77b7e69d [Translator] Update intl-messageformat range version to ^10.5.11, close #1437

weaverryan committed:
feature #1436 [Twig][Live] Drop Twig 2 support (smnandre)

This PR was squashed before being merged into the 2.x branch.

Discussion
----------

[Twig][Live] Drop Twig 2 support

As Symfony Twig Bridge / Twig Bundle have already dropped Twig 2 support... i think it's time for us to do the same thing :)

* TwigBundle:  https://github.com/symfony/twig-bundle/commit/bca6a836fdd2edad7c15278b2f11f3c7f204d395
* TwigBridge: https://github.com/symfony/twig-bridge/commit/34e9d77f152a8035cfafb27756930d898595b0c8

Commits
-------

381670a4 [Twig][Live] Drop Twig 2 support

weaverryan committed:
feature #1431 [LiveComponent] Add better error message when hydrating dates (1ed)

This PR was merged into the 2.x branch.

Discussion
----------

[LiveComponent] Add better error message when hydrating dates

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Issues        | -
| License       | MIT

Instead of

```
Return value must be of type ?object, bool returned"
```
return a proper error message
```
The model path "createdAt" was sent invalid date data "0" or in an invalid format. Make sure it's a valid date and it matches the expected format "Y. m. d.".
```

Not sure if it's a bugfix or feature. Maybe the wording can be tweaked.

Commits
-------

14fb75ec [LiveComponent] Add better error message when hydrating dates

weaverryan committed:
minor #1430 [LiveComponent] Fix component syntax error in doc (smnandre)

This PR was squashed before being merged into the 2.x branch.

Discussion
----------

[LiveComponent] Fix component syntax error in doc

Spotted by `@fGuix` in #1424 (thanks)

Commits
-------

e3a7325e [LiveComponent] Fix component syntax error in doc

weaverryan committed:
upgrading to latest ux

weaverryan committed:
minor #1530 upgrading to latest ux 2.15.0 (weaverryan)

This PR was merged into the 2.x branch.

Discussion
----------

upgrading to latest ux 2.15.0

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| Issues        | None
| License       | MIT

This will fix a few bugs that came with 2.15 on the site.

Commits
-------

59b73828 upgrading to latest ux

weaverryan committed:
upgrading mercure and specifying database options

weaverryan committed:
minor #1534 upgrading mercure and specifying database options (weaverryan)

This PR was merged into the 2.x branch.

Discussion
----------

upgrading mercure and specifying database options

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| Issues        | None
| License       | MIT

Our Mercure Bolt database is filling up - the history cleanup - https://mercure.rocks/docs/hub/config#bolt-adapter - doesn't seem to be happening, possibly due to missing options or an out-of-date mercure.

Commits
-------

61196f56 upgrading mercure and specifying database options

weaverryan committed:
[Site] Tweaking mercure command

javiereguiluz committed:
Update index.rst

kbond committed:
minor #1537 [Doc] Fix a minor syntax issue in docs (javiereguiluz)

This PR was merged into the 2.x branch.

Discussion
----------

[Doc] Fix a minor syntax issue in docs

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| Issues        | -
| License       | MIT

Commits
-------

442e2624 Update index.rst

smnandre committed:
WIP icon website

smnandre committed:
Clean CSS

smnandre committed:
Clean templates

smnandre committed:
Clean templates

smnandre committed:
Debug Modal ?!

smnandre committed:
Improve modal + tabs

smnandre committed:
Improve lot of things

smnandre committed:
Improve Modal (clipboard+tabs) + Twig templates / lists

smnandre committed:
IconSet list routes

smnandre committed:
..

smnandre committed:
Fix stuff about sets

smnandre committed:
Clean things

smnandre committed:
Fix category

smnandre committed:
Fix category

smnandre committed:
CS
